### PR TITLE
fix cut, copy, paste command handling in Emacs mode

### PR DESCRIPTION
**Pull request by kevinushey**
Not quite ready for merge, but wanted some code so we could chat about the issue + make sure I have everything straight in my head.

---

There are a number of related issues in our Emacs mode on Qt:

1. We attempt to bind some commands with a `C-c` prefix (e.g. `C-c C-v` to get help for the item under cursor), but `C-c` is caught at a very low level and turned into a copy command (before GWT even gets to see it)

2. The ShortcutManager always eats `C-c` and `C-x`, even if an Ace is not the target of the current keydown event. This breaks attempts to e.g. copy or cut from non-Ace elements in the IDE when Emacs mode is on.

3. The 'Cut', 'Copy' and 'Paste' commands don't always work when invoked from a menu, as these get translated into the corresponding keypresses (e.g. `C-x`, `C-c`, `C-v`); if Ace handles it while a Vim / Emacs keyboard handler is active, it will use those bindings in lieu of an actual copy / paste.

@jjallaire, do you recall if there's a reason why we just emit the appropriate keypresses on these commands, rather than executing the 'real' copy / cut / paste command at a lower level?

In short, the goal of this PR is to:

1. Make sure that `C-c` and `C-v` have the correct behaviour whether executed within Ace or within a separate input field, and

2. Make sure that Cut, Copy and Paste always do the right thing regardless of the active editor mode.

### CODE REVIEW

File: src/gwt/src/org/rstudio/core/client/command/ShortcutManager.java [hunk 1]
```diff
@@ -34,12 +34,15 @@
 import org.rstudio.core.client.events.NativeKeyDownHandler;
 import org.rstudio.studio.client.RStudioGinjector;
 import org.rstudio.studio.client.application.events.EventBus;
+import org.rstudio.studio.client.events.BeginCopyEvent;
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/core/client/command/ShortcutManager.java]: This should be consolidated into `BeginEditEvent`, `EndEditEvent` with enum for the edit type (cut, copy, paste)

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorWidget.java [hunk 2]
```diff
@@ -267,29 +268,12 @@ public void onAttachOrDetach(AttachEvent event)
...
-            new EndPasteEvent.Handler()
-            {
-               
-               @Override
-               public void onEndPaste(EndPasteEvent event)
-               {
-                  maybeRemapCtrlV();
-               }
-            });
+      if (!hasPasteHandlers_)
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorWidget.java]: Note that only one copy of these events need to be registered as they modify the 'global' Emacs / Vim command bindings, which are shared by all Ace instances.

**Comment by kevinushey** [comment]
Working on this also brings to light that we might want some mechanism for declaring when a particular keymap is active -- this would be a simple way of activating / deactivating groups of commands in certain scenarios. (For example, commands in an `EDITOR` keymap should only be active if an Ace editor has focus, and so on).

Will need to think about how this interacts with the dynamic command management, though.

File: src/gwt/acesupport/loader.js [hunk 3]
```diff
@@ -198,24 +198,25 @@ oop.inherits(RStudioUndoManager, UndoManager);
 
 function loadEditor(container) {
    var env = {};
+   container.env = env;
```
Comment (kevinushey) [inline src/gwt/acesupport/loader.js]: The noise below is just indentation changes -- the major change we make is ensure that the Ace editor is attached to the DOM element (so we can easily access it ourselves if needed). Note that Ace will automatically clean up these references itself when it's destroyed so we shouldn't have to worry about leaks.

**Comment by jjallaire** [comment]
@kevinushey Let's review / discuss when I am back in the office on Monday. I think the reason we emit the keypresses is to trigger some aspect of clipboard handling that's required (we wouldn't have come up with that hack otherwise). @jcheng5 Might actually recall better than me as he did the original Qt bindings and I believe also the Cocoa bindings. In general we need to treat carefully here as the existing code is all likely there for a good reason not necessarily ascertainable by quick ad-hoc testing (e.g. it might have been put in to fix a bug related to multi-byte character encodings, international keyboards, etc.)

File: src/cpp/desktop/DesktopBrowserWindow.cpp [hunk 4]
```diff
@@ -68,10 +68,6 @@ BrowserWindow::BrowserWindow(bool showToolbar,
    setUnifiedTitleAndToolBarOnMac(true);
 
    desktop::enableFullscreenMode(this, false);
-
-   QShortcut* copyShortcut = new QShortcut(QKeySequence::Copy, this);
```
Comment (kevinushey) [inline src/cpp/desktop/DesktopBrowserWindow.cpp]: Relevant commit for this here: https://github.com/rstudio/rstudio/commit/0a22b50e31f8940bf8a398f4c07701d0e6e0968a

Tested in Ubuntu Desktop and all seems okay, but probably need to think about this a bit more.

**Comment by jcheng5** [comment]
Vaguely recall there was something related to copying/pasting shortcuts working in the non-Ace textboxes, like the help topic find textbox.

File: src/cpp/desktop/DesktopBrowserWindow.cpp (hunk 4 above)
Comment (kevinushey) [inline src/cpp/desktop/DesktopBrowserWindow.cpp]: @jjallaire I've tested this fairly thoroughly and it seems like copy still works as expected, but it would be good to have some more testing post-merge. Note that removing this implies that instead of directly emitting a copy event on Ctrl + C, we instead do the dance of having Qt emit a Ctrl + C key combination, and that being handled on the client side as a copy (or other action, in the case of Emacs mode).

Would we be able to borrow some time from @dfalty to test this some more this post-merge?

File: src/gwt/src/org/rstudio/core/client/command/ShortcutManager.java [hunk 5]
```diff
@@ -337,6 +359,25 @@ public int compare(ShortcutInfo o1, ShortcutInfo o2)
    
    private boolean handleKeyDown(NativeEvent event)
    {
+      // Don't handle the (synthetic) copy, cut, paste keys
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/core/client/command/ShortcutManager.java]: Just to make it clear what's going on here, the goal is to allow e.g. Copy to succeed when executed from the menu. This is done in the following way:

- The 'copyDummy' command emits an `EditEvent(true)`, which Ace listens for and uses to suppress custom `C-c` handlers;

- A `C-c` key event is emitted from the server side (note that this is done asynchronously),

- The `ShortcutManager` receives that key event, but lets it pass through (so that it can be handled as just a regular copy). It fires a (deferred) `EditEvent(false)` event, to signal that the copy has finished and custom Ace bindings can be restored.

File: src/gwt/src/org/rstudio/core/client/command/ShortcutManager.java [hunk 6]
```diff
@@ -400,23 +461,55 @@ private boolean handleKeyDown(NativeEvent event)
...
-            return keyCode == KeyCodes.KEY_C ||
-                   keyCode == KeyCodes.KEY_X;
+            return keyCode == KeyCodes.KEY_C || keyCode == KeyCodes.KEY_X;
          }
       }
       
       return false;
    }
    
+   private static final native boolean isContentEditable(Element element) /*-{
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/core/client/command/ShortcutManager.java]: This is unused now (we should remove it)

**Comment by kevinushey** [comment]
I think this is good to go, but is still definitely a bit risky.

**Comment by jjallaire** [comment]
Given that you still think it's risky let's wait to merge for now until we
can get better comfort and/or additional review.

**Comment by kevinushey** [comment]
Bug report on the support article re: `C-c` handling in Emacs: https://support.rstudio.com/hc/en-us/articles/210928128-Emacs-Editor-Keybindings?page=1#comment_206240617.

What should we do to make ourselves confident in taking this PR? Note that this PR also fixes any Emacs keybindings with a `C-c` prefix on Qt.

**Comment by jjallaire** [comment]
Sorry it has taken us so long to merge this. There is definitely a lot going on here, @dfalty do you know how to build the IDE locally for testing prior to merge? Alternatively, Kevin, if you merge master into this branch we could cut a one-off daily build with these changes?

I'm hoping to cut a preview that we promote a bit more widely soon so I don't want to merge this before then. That said, I think testing should proceed now so we feel better about merging it when the time comes.

**Comment by dfalty** [comment]
Unfortunately no, but it's probably a good idea for me to know that. I'm guessing this will require a wide compatibility test run so it may be easier to have a one-off daily build with all platforms. @kevinushey If possible I would like to sync up on this and we can talk about the risk and scope of testing.

**Comment by jjallaire** [comment]
Okay, @kevinushey let me know once you've merged origin/master into this
branch and I'll roll a one off daily build from this branch.

**Comment by kevinushey** [comment]
Alright, this branch should be good to go. Thanks @jjallaire!

@dfalty, let me know when you want to chat about what avenues to test / check here.

**Comment by jjallaire** [comment]
Okay, I'm rolling 0.99.1167 right now which will be a new daily build from
this branch.

**Comment by dfalty** [comment]
My test pass is complete and the base functionality of cut, copy, and paste works in all scenarios. I put in a few bugs in my pass mostly related to specific Emacs use cases. 
https://rstudio.fogbugz.com/default.asp?5629
https://rstudio.fogbugz.com/default.asp?5630
https://rstudio.fogbugz.com/default.asp?5631
https://rstudio.fogbugz.com/default.asp?5632
https://rstudio.fogbugz.com/default.asp?5633
https://rstudio.fogbugz.com/default.asp?5634

**Comment by kevinushey** [comment]
It looks like these are all pre-existing bugs in the IDE (ie, none of these are newly introduced by this PR / branch); is that correct?

If so, I think that we should be safe to merge.

**Comment by dfalty** [comment]
That is correct. I went through and verified that all those bugs are not new with the branch. I agree that is safe to merge.

**Comment by jjallaire** [comment]
Let's add a NEWS item for this.

**Comment by kevinushey** [comment]
PR has a NEWS commit here: https://github.com/rstudio/rstudio/pull/589/files#diff-8312ad0561ef661716b48d09478362f3R101

(we might want to re-organize the NEWS entries a bit soon as it's starting to get quite big / cluttered!)
